class.jetpack-client-server.php

```diff
@@ -37,61 +37,45 @@ function authorize() {
 			$jetpack = $this->get_jetpack();
 			$role = $jetpack->translate_current_user_to_role();
 
-			if ( !$role ) {
-				Jetpack::state( 'error', 'no_role' );
-				break;
+			if ( ! $role ) {
+				return new Jetpack_Error( 'no_role', 'Invalid request.', 400 );
 			}
 
 			$cap = $jetpack->translate_role_to_cap( $role );
 			if ( !$cap ) {
-				Jetpack::state( 'error', 'no_cap' );
-				break;
+				return new Jetpack_Error( 'no_cap', 'Invalid request.', 400 );
 			}
 
-			$this->check_admin_referer( "jetpack-authorize_{$role}_{$redirect}" );
-
-			if ( !empty( $data['error'] ) ) {
-				Jetpack::state( 'error', $data['error'] );
-				break;
+			if ( ! empty( $data['error'] ) ) {
+				return new Jetpack_Error( $data['error'], 'Error included in the request.', 400 );
 			}
 
-			if ( empty( $data['state'] ) ) {
-				Jetpack::state( 'error', 'no_state' );
-				break;
+			if ( ! isset( $data['state'] ) ) {
+				return new Jetpack_Error( 'no_state', 'Request must include state.', 400 );
 			}
 
-			if ( !ctype_digit( $data['state'] ) ) {
-				Jetpack::state( 'error', 'invalid_state' );
-				break;
+			if ( ! ctype_digit( $data['state'] ) ) {
+				return new Jetpack_Error( $data['error'], 'State must be an integer.', 400 );
 			}
 
 			$current_user_id = get_current_user_id();
 			if ( $current_user_id != $data['state'] ) {
-				Jetpack::state( 'error', 'wrong_state' );
-				break;
+				return new Jetpack_Error( 'wrong_state', 'State does not match current user.', 400 );
 			}
 
 			if ( empty( $data['code'] ) ) {
-				Jetpack::state( 'error', 'no_code' );
-				break;
+				return new Jetpack_Error( 'no_code', 'Request must include an authorization code.', 400 );
 			}
 
 			$token = $this->get_token( $data );
 
 			if ( is_wp_error( $token ) ) {
-				if ( $error = $token->get_error_code() )
-					Jetpack::state( 'error', $error );
-				else
-					Jetpack::state( 'error', 'invalid_token' );
-
-				Jetpack::state( 'error_description', $token->get_error_message() );
-
-				break;
+				$code = $token->get_error_code() || 'invalid_token';
```

Comment:
This construct doesn't work in PHP: `false || "HELLO"` evaluates to `true` not `"HELLO"`.